network/gossip/libp2p/network.go

```diff
@@ -156,38 +171,48 @@ func (n *Network) Identity() (map[flow.Identifier]flow.Identity, error) {
 }
 
 // Topology returns the identities of a uniform subset of nodes in protocol state using the topology provided earlier
-func (n *Network) Topology() (map[flow.Identifier]flow.Identity, error) {
-	subset, err := n.top.Subset(n.ids, n.fanout())
+func (n *Network) Topology() (flow.IdentityList, error) {
+	fanout := n.fanout()
+	n.RLock()
+	defer n.RUnlock()
+	subset, err := n.top.Subset(n.ids, fanout)
```

Comment:
### ⚠️ 🕷️ **Uncovered Edge case**

I think this is an uncovered edge case:
* there are two inputs to computing the Topology: `n.ids` and `fanout`, where `fanout` also depends on the number of nodes, i.e. `len(n.ids)`
* The values are computed in _independently locked_ code blocks! For example, consider the case where we compute `fanout` but before line 178, the internal values of `n.ids` are concurrently updated. We would then compute the wrong topology.

Instead, the topology needs to be computed in a atomic operation.

Reply:
Thanks for that detailed review!
the `fanout` currently has no effect. The topology ignores it. But I agree with you, I will change this to make the whole operation atomic.